presto-main/src/main/java/com/facebook/presto/sql/gen/RowConstructorCodeGenerator.java

```diff
@@ -55,23 +61,70 @@ public BytecodeNode generateExpression(BytecodeGeneratorContext context, Type ro
                         constantInt(1))));
         block.append(singleRowBlockWriter.set(blockBuilder.invoke("beginBlockEntry", BlockBuilder.class)));
 
+        // We optimize variables.
+        Map<Type, Variable> varMap = new HashMap<>();
+        Map<Type, LabelNode> labelMap = new HashMap<>();
+        Variable objectVar = null;
+        LabelNode objectLabel = null;
         for (int i = 0; i < arguments.size(); ++i) {
             Type fieldType = types.get(i);
-            Variable field = scope.createTempVariable(fieldType.getJavaType());
-            block.comment("Clean wasNull and Generate + " + i + "-th field of row");
-            block.append(context.wasNull().set(constantFalse()));
-            block.append(context.generate(arguments.get(i), Optional.empty()));
-            block.putVariable(field);
-            block.append(new IfStatement()
-                    .condition(context.wasNull())
-                    .ifTrue(singleRowBlockWriter.invoke("appendNull", BlockBuilder.class).pop())
-                    .ifFalse(constantType(binder, fieldType).writeValue(singleRowBlockWriter, field).pop()));
+            Variable variable;
+            if (!varMap.containsKey(fieldType)) {
+                variable = scope.getSingletonVariable(fieldType.getJavaType());
+                block.initializeVariable(variable);
+                varMap.put(fieldType, variable);
+                labelMap.put(fieldType, new LabelNode("store_" + fieldType.getDisplayName()));
+            }
         }
+
+        block.append(context.wasNull().set(constantFalse()));
+        Variable index = scope.createTempVariable(int.class);
+        BytecodeBlock loopBody = new BytecodeBlock();
+        BytecodeNode forLoop = new ForLoop()
```

Comment:
Can we add comment to each block with the corresponding java code so it would be easier to read / review? Thanks!

Reply:
done